### PR TITLE
[index] Improvements on how conformances are recorded

### PR DESCRIPTION
**Pull request by akyrtzi**
Query for the local conformances of a DeclContext and record conformances of
functions that are defined in other types as 'implicit' overrides, with
the type that introduces the conformance as container symbol.

This allows more accurate index data, and avoids needing to query for all
protocol requirements, which is expensive to calculate.

### CODE REVIEW

**Comment by akyrtzi** [comment]
@swift-ci smoke test

**Comment by akyrtzi** [comment]
@swift-ci smoke test

**Review by benlangmuir — APPROVED** [review]
LGTM, with one suggestion.

File: lib/Index/Index.cpp [hunk 1]
```diff
@@ -649,12 +710,15 @@ bool IndexSwiftASTWalker::startEntityDecl(ValueDecl *D) {
       return false;
   }
 
-  for (auto Overriden: getOverriddenDecls(D, /*IncludeProtocolReqs=*/!isSystemModule)) {
-    if (addRelation(Info, (SymbolRoleSet) SymbolRole::RelationOverrideOf, Overriden))
-      return false;
+  for (auto Overriden: getOverriddenDecls(D, /*IncludeProtocolReqs=*/false)) {
```
[the remark this answers is not in the document]
Comment (akyrtzi) [reply lib/Index/Index.cpp]: Unfortunately the migrator also depends on this (in APIDiffMigratorPass::getRelatedDiffItems()), I intend to remove it after the migrator is changed as well (rdar://43844914)

**Pull request merged**